contributor-docs/updating-components.md

```diff
@@ -0,0 +1,45 @@
+# Updating components
+
+## Deprecation and removal process
+
+Our approach to making API changes and deprecations is heavily guided by a component’s usage in the [github.com codebase](https://github.com/github/github) and your confidence in the scope of your changes. If making a breaking change to a component’s API will require none or only a few updates in github.com then it is okay to release the breaking change and include any required updates in the pull request for upgrading the Primer ViewComponents library in github.com (we haven’t established a process for communicating these yet but make sure they’re detailed in your pull request).
+
+If the change requires a lot of updates or you’re unsure of the consequences of the change it’s best to follow the deprecation process:
+
+1. Make the changes without effecting existing functionality
+
+   For significant changes consider the [component lifecycle](https://primer.style/contribute/component-lifecycle) criteria; if the changes qualify the component for the next maturity level it’s much easier to introduce, for example, a new Beta component and deprecate the Alpha component than maintaining both APIs in a single component. This also makes it easier to identify which cases have been migrated.
+
+2. Mark old functionality as deprecated
+
+   For arguments add `DEPRECATED` before the YARD argument description, for components update the `status` value to `:deprecated`.
+
+3. Write an ERB linter to detect and fix uses of old functionality
+
+   This will make it easier to find uses of the old functionality in github.com. Our ERB linters also have autocorrection logic so we can automate migrations where possible. [See our linting guide for details on how to write linters](https://primer.style/view-components/linting).
+
+4. Make a release and upgrade the Primer ViewComponents library in github.com
+5. Migrate uses in github.com to new functionality
+
+   [How do we run the migrations / get data from the linter?]
+
+   If the migrations result in a lot of changes it’s best to split them up into multiple pull requests. [Is there a strategy for this to ping as few teams as possible?]
```

Comment:
@jonrohan I remember you doing this for some css vars migration. Did you use some sort of script for it? 🙏

Reply:
Yes, I did with this script https://thehub.github.com/engineering/products-and-services/dotcom/serviceowners/serviceowners-usage/#bincodeowners-stash-changes

Then I scripted that script with this to automatically publish draft prs for me:

```rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Get all currently modified files.
modified_files = `git status --short`.lines
                  .select { |line| line.start_with?(/^\s?M\s+/) }
                  .map { |line| line.strip.sub(/^M\s+/, "") }

# Get list of teams that own the modified files.
teams = `bin/codeowners #{modified_files.join(" ")}`.gsub("@github/","").split("\n")

teams.each do |team|
  full_team_name = "@github/#{team}"
  branch_name = "migrating_deprecated_system_args/#{team}"

  # checkout master
  `git checkout master`
  # stash pop
  `git stash pop`
  # create team branch
  `git checkout -b #{branch_name}`
  # run script
  `bin/codeowners-stash-changes #{full_team_name}`
  # commit changes
  `git commit -am "Migrating deprecated primer system arguments"`
  `git push -u origin #{branch_name}`
  # push draft pr
  `gh pr create -d -b "" -t "[Color modes: v2 migration] Migrating deprecated Primer View Component system arguments in github/#{team} owned files"`
end
```

Reply:
😍  a thing of beauty